File: semestr2/Lab02/PrimeNumbers.java

public class PrimeNumbers {
    int length = 0;
    int[] primes;


    public PrimeNumbers(int n) {
        boolean[] sieve = new boolean[n];
        primes = new int[n];
        for (int i = 2; i < n; i++) {
            sieve[i] = true;
        }

        for (int i = 2; i < n; i++) {
            if (sieve[i]) {
                for (int j = i + i; j < n; j = j + i) {
                    sieve[j] = false;
                }
            }
        }

        int index = 0;
        for (int i = 0; i < n; i++) {
            if (sieve[i]) {
                primes[index] = i;
                length++;
            }
            index++;
        }
    }

    public int getNumber(int m) {
        return primes[m];
    }

    public int getLength() {
        return length;
    }
}

class Test {
    public static void main(String[] args) {
        try {
            int n = Integer.parseInt(args[0]);
            if (n <= 1) {
                System.out.println(n + " - " + " Nieprawidłowy zakres");
                return;
            }
            PrimeNumbers primes = new PrimeNumbers(n);
            for (int i = 1; i < args.length; i++) {
                try {
                    int currentInteger = Integer.parseInt(args[i]);
                    if (currentInteger >= primes.getLength()) {
                        System.out.println(args[i] + " - " + " liczba spoza zakresu");
                    } else {
                        System.out.println(args[i] + " - " + primes.getNumber(currentInteger));
                    }
                } catch (NumberFormatException ex) {
                    System.out.println(args[i] + " nieprawidłowa dana");
                }

            }
        } catch (NumberFormatException ex) {
            System.out.println(args[0] + " nieprawidłowa dana");
            return;
        }

    }
}

File: semestr2/Lab02/C++/PrimeNumbers.cpp
#include "PrimeNumbers.hpp"
#include <iostream>

PrimeNumbers::PrimeNumbers(int n)
{
    bool *sieve = new bool[n + 1];
    primes = new int[n + 1];
    for (int i = 0; i <= n; i++)
    {
        sieve[i] = true;
    }
    sieve[0] = sieve[1] = false;

    // sprawdzanie do sqrt(n)
    for (int i = 2; i * i <= n; i++)
    {
        if (sieve[i])
        {
            // zaczynamy od i*i bo wszystkie i+i przed i*i zostały już sprawdzone przez mniejsze dzielniki  
            for (int j = i * i; j <= n; j = j + i)
            {
                sieve[j] = false;
            }
        }
    }

    for (int i = 0; i <= n; i++)
    {
        if (sieve[i])
        {

            primes[length] = i;
            length++;
        }
    }
    delete[] sieve;
}

int PrimeNumbers::getNumber(int m)
{
    return primes[m];
}

int PrimeNumbers::getLength()
{
    return length;
}

PrimeNumbers::~PrimeNumbers()
{
    delete[] primes;
}

int main(int argc, char *argv[])
{
    if (argc == 1)
    {
        std::cout << "Nie podano argumentów";
        return 0;
    }
    try
    {
        int n = std::stoi(argv[1]);
        if (n <= 1)
        {
            std::cout << n << " - " << " nieprawidłowy zakres\n";
            return 0;
        }
        // PrimeNumbers primes(n);
        PrimeNumbers *primes = new PrimeNumbers(n);
        for (int i = 2; i < argc; i++)
        {
            try
            {
                int currentInteger = std::stoi(argv[i]);
                if (currentInteger >= primes->getLength() || currentInteger < 0)
                {
                    std::cout << argv[i] << " - " << " liczba spoza zakresu\n";
                }
                else
                {
                    std::cout << argv[i] << " - " << primes->getNumber(currentInteger) << "\n";
                }
            }
            catch (...)
            {
                std::cout << argv[i] << " nieprawidłowa dana\n";
            }
        }
        delete primes;
    }
    catch (...)
    {
        std::cout << argv[1] << " nieprawidłowa dana\n";
        return 0;
    }
}